### PR TITLE
Give a keyboard shortcut to switch default project #653

### LINKED ISSUE #653: Give a keyboard shortcut to switch default project
as per title

### PR DESCRIPTION
**Pull request by Honoo**
Resolves #653. The keyboard shortcut is `Ctrl` + `R`. When it is pressed, the default repository switches to the next open repository in the set of open repositories.

### CODE REVIEW

**Comment by acruis** [comment]
Failed coding style checks. Do indent properly :)

File: src/main/java/ui/UI.java [hunk 1]
```diff
@@ -415,6 +420,26 @@ private void primaryRepoChanged(String repoId) {
...
+    	String[] openRepos = logic.getOpenRepositories().toArray(new String[0]);
+    	String currentRepo = logic.getDefaultRepo();
+    	
+    	// Cycle to the next open repository
+    	for(int i = 0; i < openRepos.length; i++){
+    		if(openRepos[i].equals(currentRepo)){
+    			if(i == openRepos.length-1){
+    				primaryRepoChanged(openRepos[0]);
+    			}
+    			else {
```
Comment (damithc) [inline src/main/java/ui/UI.java]: `else` should be in the same line as `}`. Follow the coding standard.

**Comment by damithc** [comment]
Remember to add tests too.

**Comment by lejolly** [comment]
Yeah the ci build log caught quite a few coding style errors, please check and fix them.

**Comment by Honoo** [comment]
Ok, is there a style guide I could check?

**Comment by lejolly** [comment]
I think it should be similar to the CS2103 one, if not use the [TEAMMATES](https://docs.google.com/document/pub?id=1iAESIXM0zSxEa5OY7dFURam_SgLiSMhPQtU0drQagrs&embedded=true) one. 

Yeah we run individual tests using JUnit and the entire test suite using gradle (on windows it would be `gradlew.bat clean test`). Could you attach the stacktrace of the NPE?

**Comment by Honoo** [comment]
This is the stack trace when I run ValidLoginTest:

`Exception in Application start method
Exception in thread "Thread-2" java.lang.RuntimeException: Exception in Application start method
	at com.sun.javafx.application.LauncherImpl.launchApplication1(Unknown Source)
	at com.sun.javafx.application.LauncherImpl.lambda$launchApplication$152(Unknown Source)
	at com.sun.javafx.application.LauncherImpl$$Lambda$7/1524636.run(Unknown Source)
	at java.lang.Thread.run(Unknown Source)
Caused by: java.lang.NullPointerException
	at ui.UI.initCSS(UI.java:247)
	at ui.UI.initUI(UI.java:180)
	at ui.UI.start(UI.java:88)
	at guitests.UITest$TestUI.start(UITest.java:26)
	at com.sun.javafx.application.LauncherImpl.lambda$launchApplication1$159(Unknown Source)
	at com.sun.javafx.application.LauncherImpl$$Lambda$57/15764541.run(Unknown Source)
	at com.sun.javafx.application.PlatformImpl.lambda$runAndWait$172(Unknown Source)
	at com.sun.javafx.application.PlatformImpl$$Lambda$53/8480533.run(Unknown Source)
	at com.sun.javafx.application.PlatformImpl.lambda$null$170(Unknown Source)
	at com.sun.javafx.application.PlatformImpl$$Lambda$55/13211503.run(Unknown Source)
	at java.security.AccessController.doPrivileged(Native Method)
	at com.sun.javafx.application.PlatformImpl.lambda$runLater$171(Unknown Source)
	at com.sun.javafx.application.PlatformImpl$$Lambda$54/5606483.run(Unknown Source)
	at com.sun.glass.ui.InvokeLaterDispatcher$Future.run(Unknown Source)
	at com.sun.glass.ui.win.WinApplication._runLoop(Native Method)
	at com.sun.glass.ui.win.WinApplication.lambda$null$145(Unknown Source)
	at com.sun.glass.ui.win.WinApplication$$Lambda$43/24933202.run(Unknown Source)
	... 1 more`

**Comment by lejolly** [comment]
looks like a test resources problem, did you define the test resource folder in the project? `src/test/resources`

**Comment by Honoo** [comment]
Where should I define the test resource folder?

**Comment by lejolly** [comment]
Try [this](http://stackoverflow.com/a/2912652) if you're using eclipse? Can't guarantee that it'll work since I don't use it. 

For IntelliJ, this is what it looks like for me:
project_structure

**Comment by Honoo** [comment]
The StackOverflow link worked, thanks!

**Comment by lejolly** [comment]
could you update this to work with the latest changes in master?

File: src/main/java/ui/UI.java [hunk 2]
```diff
@@ -443,6 +448,27 @@ private void primaryRepoChanged(String repoId) {
         logic.openPrimaryRepository(repoId);
         logic.setDefaultRepo(repoId);
     }
+    
+    public void switchDefaultRepo(){
+        logger.info("Triggered switch default repo");
+        String[] openRepos = logic.getOpenRepositories().toArray(new String[0]);
```
Comment (acruis) [inline src/main/java/ui/UI.java]: `openRepos` might not be sorted as the sorting is done in RepositorySelector. Why not just get the contents from the RepositorySelector's combo box?

File: src/main/java/ui/UI.java [hunk 3]
```diff
@@ -443,6 +448,27 @@ private void primaryRepoChanged(String repoId) {
...
+        String currentRepo = logic.getDefaultRepo();
+        
+        // Cycle to the next open repository
+        for (int i = 0; i < openRepos.length; i++) {
+            if (openRepos[i].equals(currentRepo)) {
+                if (i == openRepos.length - 1) {
+                    primaryRepoChanged(openRepos[0]);
+                    UI.events.triggerEvent(new DefaultRepoSwitchedEvent(openRepos[0]));
```
Comment (acruis) [inline src/main/java/ui/UI.java]: Try to limit the use of test events.

File: src/main/java/ui/UI.java (hunk 2 above)
Comment (Honoo) [inline src/main/java/ui/UI.java]: But the combo box only allows you to get one open repository using `getText`, right?

Comment (acruis) [inline src/main/java/ui/UI.java]: `RepositorySelector` already accesses the combo box's items through `getItems()` in its `loadContents` method. Try using `getItems()` to see if you can get the full repo list?

**Comment by Honoo** [comment]
May I know if this is ready to to be merged?

**Comment by lejolly** [comment]
if `DefaultRepoSwitchedEvent` is only used in tests please put it in the `testevents` package.

**Comment by acruis** [comment]
Actually, we have a `PrimaryRepoChangedEvent` already. Why not use that?

**Comment by Honoo** [comment]
Thanks for the suggestions, I've incorporated them.

**Comment by acruis** [comment]
OK, good to go.

File: src/main/java/ui/UI.java [hunk 4]
```diff
@@ -443,6 +448,24 @@ private void primaryRepoChanged(String repoId) {
...
+        // Cycle to the next open repository
+        for (int i = 0; i < openRepos.length; i++) {
+            if (openRepos[i].equals(currentRepo)) {
+                if (i == openRepos.length - 1) {
+                    primaryRepoChanged(openRepos[0]);
+                    repoSelector.setText(openRepos[0]);
+                } else {
+                    primaryRepoChanged(openRepos[i + 1]);
+                    repoSelector.setText(openRepos[i + 1]);
+                }
```
Comment (damithc) [inline src/main/java/ui/UI.java]: This is more readable:
```
boolean isLastInTheList = (i == openRepos.length - 1);
int nextRepoIndex = (isLastInTheList ? 0 : i+1);
primaryRepoChanged(nextRepoIndex);
repoSelector.setText(nextRepoIndex);             
```

Or this:
```
//go back to first item if current selected item is the last one in the list
int nextRepoIndex = (i == openRepos.length - 1 ? 0 : i+1);
primaryRepoChanged(nextRepoIndex);
repoSelector.setText(nextRepoIndex);             
```

File: src/test/java/guitests/UIEventTests.java [hunk 5]
```diff
@@ -66,4 +75,27 @@ public void panelClickedTest() {
...
+        doubleClick(comboBox);
+        doubleClick();
+        type("dummy3/dummy3");
+        push(KeyCode.ENTER);
+        click("#dummy/dummy_col0_filterTextField");
+        resetEventTestCount();
+        press(KeyCode.CONTROL).press(KeyCode.R).release(KeyCode.R).release(KeyCode.CONTROL);
+        assertEquals(1, eventTestCount);
+        press(KeyCode.CONTROL).press(KeyCode.R).release(KeyCode.R).release(KeyCode.CONTROL);
+        assertEquals("dummy3/dummy3", defaultRepoId);
```
Comment (damithc) [inline src/test/java/guitests/UIEventTests.java]: Did you test for boundary cases of repo count 0, 1?

Comment (acruis) [inline src/test/java/guitests/UIEventTests.java]: Repo count 0 is not really valid to test for i.e. we don't have a mechanism to start the UI with zero repos opened. Repo count 1 is a valid case, though.

File: src/main/java/ui/UI.java (hunk 4 above)
Comment (acruis) [inline src/main/java/ui/UI.java]: That's neater. I'll make another issue for the dropdown, too.
